unsupervised_rock_detection_3d/README.md

```diff
@@ -4,10 +4,52 @@
+
+<img src="./images/Original.png" height="200"> <img src="./images/RegionGrowing2.png" height="200">
+
 ## Workflow
 
 The module involves two primary unsupervised segmentation techniques:
 
+### Region Growing Segmentation
+
+1. **Initialization**:
+
+    * Configure the RegionGrowingSegmentation class with the point cloud and segmentation parameters such as voxel size, number of neighbors, smoothness threshold, distance threshold, and curvature threshold.
+    * Downsample the point cloud using a voxel grid to reduce computation.
+    * Estimate normals for the downsampled point cloud to use in segmentation criteria. Normals are oriented consistently for reliable dot product calculations.
+    * Build a KD-Tree for efficient neighbor searches.
+
+2. **Seed Selection**:
+
+    * Compute the bounding box of the point cloud to find the geometric bounds.
+    * Determine the centroid of the bounding box in the x and y dimensions.
+    * Identify the point with the highest z value near the centroid, representing a potential rock seed.
+    * Identify the lowest point in the bounding box, representing a potential terrain seed.
+
+3. **Region Growing**:
+
+    * Precompute neighbors for each point within the specified distance threshold to speed up segmentation.
+    * Initialize regions from the selected seed points (highest point for rocks, bottommost point for terrain).
+    * Expand regions from seeds by evaluating neighbors based on the segmentation criteria (normal vector smoothness and curvature):
+        * Start with the seed point in a queue.
+        * For each point, evaluate its neighbors based on the segmentation criteria.
+        * Check if neighbors meet the smoothness and curvature thresholds. If they do, add them to the current region and the queue.
+        * Assign labels to points based on the region they belong to.
+4. **Post Processing**:
+
+    * Perform conditional label propagation to assign labels to remaining unlabeled points based on majority labels of their neighbors.
+    * Color the segmented point cloud for visualization by assigning different colors to different regions (e.g., red for rocks, blue for terrain).
+    * Visualize the segmented point cloud to assess the regions' boundaries.
+
+#### Novelty
+
+One significant novelty in our approach is the use of second-order neighbors for smoothness calculation. Our point cloud is densely packed, resulting in neighboring points being very close to one another. This proximity causes the normal orientation differences to appear nearly parallel, leading almost every point to be included in the region. To address this issue, we modified the approach by considering the minimum dot product (maximum deviation) with second-order neighboring points instead of comparing the immediate neighboring points. This adjustment improves the segmentation accuracy by preventing the over-inclusion of points in a region due to near-parallel normals in densely packed clouds.
```

Comment:
you should mention the motivation of using second-order neighbors in region growing

Reply:
I've already discussed the issue we faced and the reason behind using this approach. Is there anything else specific you think should be included or any specific format?

Reply:
Ah I didn't read carefully. Thanks for clarification. We are good